src/Types/BigNum.js

```diff
@@ -13,16 +13,57 @@ exports.zero = lib.BigNum.zero();
 
 exports.one = lib.BigNum.one();
 
+const add = (a, b) => {
+  return a + b;
+};
+
+const mul = (a, b) => {
+  return a * b;
+};
+
+const check_limit_with_func = (a, b, limit, func) => {
+  const r = func(a, b);
+  if (r < limit) {
+    return r;
+  } else {
+    throw new Error("Overflow detected");
+  }
+};
+
+const check_limit = (str, limit) => {
+  if (0 <= BigInt(str) && BigInt(str) < limit) {
+    return BigInt(str);
+  } else {
+    throw new Error("Overflow detected");
+  }
+};
+
 exports.bnAdd = maybe => lhs => rhs => {
+  // this is needed because try/catch overuse breaks runtime badly
+  // https://github.com/Plutonomicon/cardano-transaction-lib/issues/875
   try {
+    check_limit_with_func(
+      lhs.to_str(),
+      rhs.to_str(),
+      BigInt("18446744073709551616"),
+      add
+    ); // 2 ^ 64
     return maybe.just(lhs.checked_add(rhs));
   } catch (_) {
     return maybe.nothing;
   }
 };
 
 exports.bnMul = maybe => lhs => rhs => {
+  // this is needed because try/catch overuse breaks runtime badly
+  // https://github.com/Plutonomicon/cardano-transaction-lib/issues/875
   try {
+    check_limit_with_func(
+      lhs.to_str(),
+      rhs.to_str(),
+      BigInt("18446744073709551616"),
+      mul
+    ); // 2 ^ 64
```

Comment:
Here, the arguments to `mul` are given as strings, but `mul` is using the built-in `*` js operator.

Perhaps this fixes:
```suggestion
    check_limit_with_func(
      BigInt(lhs.to_str()),
      BigInt(rhs.to_str()),
      BigInt("18446744073709551616"),
      mul
    ); // 2 ^ 64
```

Reply:
Same with the `add` version. I also don't think we need this `check_limit_with_func`, just apply the function directly?